### PR TITLE
Generate resource key from the resource name

### PR DESCRIPTION
**Pull request by bethanyaconnor**
This was an idea that [came out of the internal bug bash](https://docs.google.com/document/d/1LTP1_UwCtJ8n_ju5qspeBSYgrh-t7MHCwKGboPKSg5A/edit?disco=AAAAKlnxy_o). Instead of having users generate a key, do it automatically based on the name. This is loosely based on the logic we have to [generate usernames](https://github.com/code-dot-org/code-dot-org/blob/6092ee3ce9e5d87834c4e3a679ed3fdbec22b5d9/lib/cdo/user_helpers.rb#L10), though I removed some of the constraints as we're significantly less likely to have many collisions. See the tests for some examples, but the logic replaces non-alphanumeric characters (plus underscores and periods) with underscores. Then we check for collisions and if there is one, start appending numbers.

I removed the `Embed Slug` field from the dialog, but the slug still appears in the table. Not sure if that's confusing to the end user or not, so would love to hear other's thoughts!

## Links

- [spec]()
- [jira]()

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by davidsbailey** [comment]
I like where this is going, but i have a concern about the following user workflow:
1. create a resource
2. try to add it to my markdown via the work being done in [PLAT-416](https://codedotorg.atlassian.net/browse/PLAT-416)

will the resource key be available in that case? 

if not, it looks like you are sending the attributes back to the client in response to the create call on the server, so maybe there is a way to update the resource on the client with the new key? I may be missing something about how this all works (maybe PLAT-416 makes its own call to the server), so feel free to ignore me if this is already taken care of.

File: dashboard/app/models/resource.rb [hunk 1]
```diff
@@ -51,4 +58,20 @@ def summarize_for_lesson_plan
       type: type
     }
   end
+
+  private
+
+  def generate_key_from_name
+    key_prefix = name.gsub(/[^a-z0-9\-\_\.]+/, '_')
+    potential_clashes = course_version_id ? Resource.where(course_version_id: course_version_id) : Resource.all
+    potential_clashes = potential_clashes.where("resources.key like '#{key_prefix}%'").pluck(:key)
```
Comment (davidsbailey) [inline dashboard/app/models/resource.rb:67]: in order for this to be an indexed query, we will need to include the course_version_id. I think this is fine to do later, once we are always only storing / looking within a course version, but wanted to call it out now while I'm thinking about it.

File: apps/test/unit/lib/levelbuilder/lesson-editor/AddResourceDialogTest.jsx [hunk 2]
```diff
@@ -21,18 +21,14 @@ describe('AddResourceDialog', () => {
   it('renders default props', () => {
     const wrapper = shallow(<AddResourceDialog {...defaultProps} />);
     expect(wrapper.contains('Add Resource')).to.be.true;
-    expect(wrapper.find('input').length).to.equal(7);
+    expect(wrapper.find('input').length).to.equal(6);
```
Comment (davidsbailey) [inline apps/test/unit/lib/levelbuilder/lesson-editor/AddResourceDialogTest.jsx:24]: just curious, why did this change?

**Review by davidsbailey — APPROVED** [review]
LGTM! I think the top level comment can be addressed in another PR, and may be hard to do before then anyway.

File: apps/test/unit/lib/levelbuilder/lesson-editor/AddResourceDialogTest.jsx (hunk 2 above)
Comment (davidsbailey) [reply apps/test/unit/lib/levelbuilder/lesson-editor/AddResourceDialogTest.jsx:24]: oh, sorry, I see this is counting inputs, not resources. you can ignore me here

File: apps/src/lib/levelbuilder/lesson-editor/AddResourceDialog.jsx [hunk 3]
```diff
@@ -94,21 +93,18 @@ export default class AddResourceDialog extends Component {
...
-    }
-    if (key === '') {
-      error += 'Embed slug is required.';
+      error += 'URL is required.';
     }
+    this.setState({error});
     if (error === '') {
       return true;
     } else {
-      this.setState({error});
```
Comment (dmcavoy) [inline apps/src/lib/levelbuilder/lesson-editor/AddResourceDialog.jsx:111]: Do you still need to set the state here?

**Review by dmcavoy — APPROVED** [review]
This is awesome Bethany! I think still showing the key in the table is worth it because that will be how users can reference it in markdown

**Comment by dmcavoy** [comment]
Also just wanted to echo Dave's wonders about how to works with the resource links we were working on together yesterday. We were counting on the key information being available in redux. Can we still get that?

File: dashboard/app/models/resource.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/models/resource.rb:67]: That makes sense! I have a [ticket](https://codedotorg.atlassian.net/browse/PLAT-434) to fix the course version/resource relationships and will keep this in mind then

**Comment by bethanyaconnor** [comment]
To answer the question about whether the key will be available in redux, I believe yes. The generated key is passed back down to the client after the form submit.

Sorry for not including screenshots or a gif earlier. You can sort of see that there's a key in the table for the new resource, which matches the one on the server:
resource-with-generated-key

**Comment by davidsbailey** [comment]
Looks awesome, Bethany! One gripe -- it looks like resource keys are generally generated with dashes, rather than underscores. I think dashes may also be slightly better in urls. So, I would suggest switching to dashes, though I am open to pushback.

**Comment by bethanyaconnor** [comment]
That seems fine to me. I can definitely imagine dashes feeling a bit more natural as well.

I'll do that in a follow-up PR though. I'd like to merge this PR soon so that I can handle merge conflicts with #37540, as well as confirming that all this does indeed work with the new redux store.

**Comment by davidsbailey** [comment]
👍 🚢 🚀

**Pull request merged**